### ISSUE TITLE
Multiple cursors binding

### ISSUE THREAD

**Issue by berdario** [Microsoft/vscode#3091]
On most Linux desktops `alt+click` will allow you drag a window by clicking anywhere over it. This conflicts with the multiple cursors binding.

It's still possible to obtain multiple cursors with commands like `insertCursorAbove` or `insertCursorAtEndOfEachLineSelected`, but this affects multiple people, and it's already covered in some [articles](https://2buntu.com/articles/1529/visual-studio-code-comes-to-linux/#multiple-cursors_3044) with a workaround (which I'm copying here for Gnome based systems: `gsettings set org.gnome.desktop.wm.preferences mouse-button-modifier "<Super>"` change `<Super>` with your preferred key )

It's also acknowledged [in VScode's documentation](https://code.visualstudio.com/Docs/editor/editingevolved#_selection-multicursor) that this shortcut might conflict even on Windows or Macosx...

> Note: Your graphics card provider might overwrite these default shortcuts.

So, fixing this for Linux users might also be useful elsewhere. For comparison, the Atom editor uses `ctrl+click` instead of `alt+click` for this shortcut.

Rather than changing the default, making it customizable would already be a welcome improvement, but afaik unfortunately there's no way to change it. These are a couple of related default keybindings:

```
{ "key": "ctrl+shift+up",         "command": "editor.action.insertCursorAbove",
                                 "when": "editorTextFocus" },
{ "key": "shift+alt+up",          "command": "editor.action.insertCursorAbove",
                                 "when": "editorTextFocus" }
```

But since I cannot find any `editor.action.insertCursor` command, it seems that this option is lacking.

**Comment by Tyriar** [comment]
This is fairly easy to change to <kbd>ctrl</kbd> in `MouseHandler._updateMouse`, however then it conflicts with <kbd>ctrl</kbd>+click to open a link. Atom doesn't seem to support <kbd>ctrl</kbd>+click to open links so there is no conflict there. @berdario <kbd>ctrl</kbd>+<kbd>alt</kbd> and <kbd>alt</kbd>+<kbd>shift</kbd> seem to be free, just adding another modifier may be better than changing both add a selection and open a link for Linux.

I don't think we currently allow customization of mouse shortcuts, only pure keyboard ones. Created https://github.com/Microsoft/vscode/issues/3130 for this.

**Comment by Tyriar** [comment]
I'm leaning towards <kbd>alt</kbd>+<kbd>shift</kbd> so it matches the default for add selection up/down.

**Comment by Tyriar** [comment]
@alexandrudima What do you think we should make the <kbd>alt</kbd>+click shortcut on Linux? <kbd>alt</kbd>+click can't be used because it conflicts on the OS-level. I was thinking <kbd>alt</kbd>+<kbd>shift</kbd>+click previously but it seems that's used to expand the selection of the last selection without touching the others.

Atom uses <kbd>ctrl</kbd>+click, but that would conflict with link clicking in VSCode. Would adding <kbd>ctrl</kbd> to the shortcuts such that <kbd>alt</kbd>+<kbd>ctrl</kbd>+click adds a cursor and <kbd>alt</kbd>+<kbd>ctrl</kbd>+<kbd>shift</kbd>+click expands the last cursor selection work?

**Comment by Reanmachine** [comment]
@Tyriar I did some testing, <kbd>ctrl</kbd>+<kbd>alt</kbd> will work, it does not trigger the same window move behavior in gnome that <kbd>alt</kbd>+click does, <kbd>shift</kbd>+<kbd>alt</kbd> does still trigger the move mode.

Whichever combination you end up going with can we please get those configurable in the `keybindings.json` file?

**Comment by alexandrudima** [comment]
The mouse gesture dispatching doesn't go through the keybindings rules, but I agree this should be configurable somehow.

Here is where the built-in editor mouse dispatching happens: https://github.com/Microsoft/vscode/blob/c4b514244e2651117b9990295730265f5b902436/src/vs/editor/browser/view/viewController.ts#L66

There are other places in code, such as the link handler that handle Ctrl+Click programmatically.

It would be a bit tricky, but doable I think, to have the `IMouseDispatchData` go through a rule engine that determines what command to invoke on the editor.

We can easily change the code to do something different in Linux, but implementing a rule engine (that would ultimately solve all problems since it would allow configuration) would be a bit more complicated.

**Comment by Reanmachine** [comment]
@alexandrudima I feel the rule engine is worth it, fixing the conflict with <kbd>alt</kbd>+Click is really a short sighed solution. As I'm sure you're aware linux can be quite varied (read: fragmented) sometimes. Fixing the conflict will solve issues for people running Gnome, but it may end up conflicting with KDE, or with XFCE, or with whatever other variant someone has decided to run. I think it's best to go the rule engine route, even if it takes a little longer.

Also, there's no telling what Apple or even Microsoft may do with mouse/touch gestures in the future, it's an evolving UX space, I think it would be better than playing cat and mouse with hardcoded gesture bindings.

**Comment by Tyriar** [comment]
@Reanmachine totally agree, the customizable mouse shortcuts is captured in this issue https://github.com/Microsoft/vscode/issues/3130.

**Comment by kirtangajjar** [comment]
Any updates on this guys?

**Comment by Tyriar** [comment]
This is blocked on https://github.com/Microsoft/vscode/issues/3130 imo

**Comment by 1415926535** [comment]
This fixed my xenial.
`gsettings set org.gnome.desktop.wm.preferences mouse-button-modifier "<Super>"`

**Comment by alexandrudima** [comment]
Fixed via #27193 

There is now a configuration option `editor.multicursorModifier` that can be set to `ctrl` or `alt` on Linux. The go to definition and open link modifiers will adjust accordingly.

**Comment by alexandrudima** [comment]
Further changes were done in #27680 -- the setting is now `editor.multiCursorModifier` which can be `ctrlCmd` or `alt`

**Comment by bpasero** [comment]
Marking as verified because this was on the test plan.

**Comment by Ratismal** [comment]
Is there no way to specify a completely custom modifier? I'm too used to using `ctrl/cmd` for the goto to switch, but at the same time being able to set multiple cursors anywhere is too big for me to lose. I'd ideally like to set it to a combination of keys, such as `ctrl+alt`.

**Comment by alexandrudima** [comment]
@Ratismal Unfortunately not at this time. We have an item #3130 that tracks that. It is, however, a bit tricky to express all the various imperative code we use today for mouse operations in a declarative format that is also straight-forward enough that end-users could edit.

**Comment by Ratismal** [comment]
Fair enough @alexandrudima. I just unbound the alt+click shortcut from my xfce settings, since I never used it anyways and it was actually interfering with other programs.

Thanks for the speedy response!

**Comment by luncht1me** [comment]
@Ratismal Just use the editor config that alex noted.
`"editor.multiCursorModifier": "ctrlCmd",`
in your User Settings will allow you to Ctrl + Click for multi cursor. It works great!

**Comment by Ratismal** [comment]
@luncht1me Yes, but as mentioned I'd lose the ability to use the goto feature. It's far better to just unbind the global alt+click hotkey so that my workspace on linux and windows is the same.
